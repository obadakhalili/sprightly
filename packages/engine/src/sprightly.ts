export = function sprightly() {
  console.log("sprightly")
}
