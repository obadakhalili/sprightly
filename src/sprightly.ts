import { readFileSync } from "fs"
import path from "path"

interface Data {
  [key: string]: string | number | Array<string | number | Data> | Data
}

class KeyNotFoundError extends Error {
  readonly key: string
  readonly src: string
  constructor(key: string, src: string) {
    const message = `Key "${key}" not found in "data" at ${src}`
    super(message)
    this.key = key
    this.src = src
  }
}

class PathNotFoundError extends Error {
  readonly component: string
  readonly src: string
  constructor(component: string, src: string) {
    const message = `Component "${component}" was not found in ${src}`
    super(message)
    this.component = component
    this.src = src
  }
}

interface Options {
  keyFallback?: string
  throwOnKeyNotfound?: boolean
  cache?: boolean
}

function parse(file: string, data: Data, options: Options) {
  const parsedFile = file.replaceAll(
    /\{\{(>?)(.*?)\}\}/g,
    (_, isComponent: string, reference: string) => {
      reference = reference.trim()
      if (isComponent) {
        try {
          return sprightly(reference, data, options)
        } catch (e) {
          throw new PathNotFoundError(reference, "")
        }
      }

      return reference
        .split(/((?:\w+))(?:(?:\[)(\d+)(?:\]))?(?:\.?)/gm)
        .filter((v) => v?.trim())
        .reduce((obj, key) => {
          key = key.trim()
          if (
            typeof obj === "object" &&
            options.throwOnKeyNotfound &&
            !(key in obj)
          ) {
            throw new KeyNotFoundError(key, "")
          }

          return (obj[key] || options.keyFallback) as Data
        }, data) as unknown as string
    },
  )
  return parsedFile
}

const cache = new Map<string, string>()

const defaultOptions = {
  keyFallback: "",
  throwOnKeyNotfound: false,
  cache: false,
}

function sprightly(
  entryPoint: string,
  data: Data,
  options: Options = defaultOptions,
): string {
  options = {
    ...defaultOptions,
    ...options,
  }

  console.log(path.resolve(entryPoint))
  const entryPointAbsolutePath = path.resolve(path.normalize(entryPoint))
  if (options.cache && cache.has(entryPointAbsolutePath)) {
    return cache.get(entryPointAbsolutePath)!
  }

  try {
    const file = readFileSync(entryPointAbsolutePath).toString()
    const parsedFile = parse(file, data, options)

    if (options.cache && !cache.has(entryPointAbsolutePath)) {
      cache.set(entryPointAbsolutePath, parsedFile)
    }

    return parsedFile
  } catch (error) {
    if (error instanceof KeyNotFoundError) {
      throw new KeyNotFoundError(error.key, entryPoint)
    }
    if (error instanceof PathNotFoundError) {
      throw new PathNotFoundError(error.component, entryPoint)
    }
    throw error
  }
}
export = function sprightlyAsync(
  entryPoint: string,
  data: Data,
  options: Options = defaultOptions,
) {
  return new Promise<string>((resolve, reject) => {
    try {
      resolve(sprightly(entryPoint, data, options))
    } catch (error) {
      reject(error as string)
    }
  })
}
