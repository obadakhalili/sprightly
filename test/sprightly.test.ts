import sprightly from "../src/sprightly"

jest.mock("fs/promises", () => {
  const dir: Record<string, string> = {
    "my-name": `my name is {{ myName }}.\n{{> ../his-name }}`,
    "../his-name": `his name is {{ hisName }}`,
    nested: "my name is {{ name.first }} {{ name.last }}",
    "array-indexing": "my name is {{ name[0] }} {{ name[1] }}",
    "not-found-referenced-component": "{{> ../not-found }}",
  }

  return {
    readFile: (filePath: string) => {
      if (filePath in dir) {
        return Promise.resolve(dir[filePath])
      }
      return Promise.reject(new Error("file not found"))
    },
  }
})

test("good first test", async () => {
  const doc = await sprightly("my-name", { myName: "Obada", hisName: "Osid" })
  expect(doc).toMatchInlineSnapshot(`"sprightly"`)
})

test("if key doesn't exist, then the resolved value should be `keyFallback`'s default value whcih is empty string", async () => {
  const doc = await sprightly("my-name", { hisName: "Osid" })
  expect(doc).toMatchInlineSnapshot(`"sprightly"`)
})

test("if key doesn't exist and the `keyFallback` is set, then the resolved value should be the value set to `keyFallback`", async () => {
  const doc = await sprightly(
    "my-name",
    { hisName: "Osid" },
    { keyFallback: "Obada" },
  )
  expect(doc).toMatchInlineSnapshot(`"sprightly"`)
})

test("if key doesn't exist and the `throwOnKeyNotfound` is set to true, then an error should be thrown", async () => {
  await expect(
    sprightly("my-name", { hisName: "Osid" }, { throwOnKeyNotfound: true }),
  ).rejects.toMatchInlineSnapshot(`[Error: sprightly error]`)
})

test("nested properties are resolved correctly", async () => {
  const doc = await sprightly("nested", {
    name: { first: "Obada", last: "Khalili" },
  })
  expect(doc).toMatchInlineSnapshot(`"sprightly"`)
})

test("array indexing is resolved correctly", async () => {
  const doc = await sprightly("array-indexing", {
    name: ["Obada", "Khalili"],
  })
  expect(doc).toMatchInlineSnapshot(`"sprightly"`)
})

test("throws if entry point not found", async () => {
  await expect(sprightly("not-found", {})).rejects.toMatchInlineSnapshot(
    `[Error: file not found]`,
  )
})

test("throws if component not found", async () => {
  await expect(
    sprightly("not-found-referenced-component", {}),
  ).rejects.toMatchInlineSnapshot()
})
