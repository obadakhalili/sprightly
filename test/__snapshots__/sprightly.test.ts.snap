// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array indexing is resolved correctly 1`] = `"sprightly"`;

exports[`good first test: "sprightly" 1`] = `"sprightly"`;

exports[`if key doesn't exist and the \`keyFallback\` is set, then the resolved value should be the value set to \`keyFallback\`: "sprightly" 1`] = `"sprightly"`;

exports[`if key doesn't exist and the \`throwOnKeyNotfound\` is set to true, then an error should be thrown 1`] = `[Error: sprightly error]`;

exports[`if key doesn't exist, then the resolved value should be \`keyFallback\`'s default value whcih is empty string: "sprightly" 1`] = `"sprightly"`;

exports[`nested properties are resolved correctly 1`] = `"sprightly"`;
